biz.aQute.bndlib/src/aQute/bnd/component/DSAnnotations.java

```diff
@@ -26,55 +50,150 @@ public boolean analyzeJar(Analyzer analyzer) throws Exception {
 		if (sc != null && sc.trim().length() > 0)
 			names.add(sc);
 
+		TreeSet<String> provides = new TreeSet<String>();
+		TreeSet<String> requires = new TreeSet<String>();
+		Version packageVersion = getPackageVersion(analyzer);
+		Version maxVersion = AnnotationReader.V1_0;
+
 		for (Clazz c : list) {
 			for (Instruction instruction : instructions.keySet()) {
 
 				if (instruction.matches(c.getFQN())) {
 					if (instruction.isNegated())
 						break;
-					ComponentDef definition = AnnotationReader.getDefinition(c, analyzer);
+					ComponentDef definition = AnnotationReader.getDefinition(c, analyzer, flags);
 					if (definition != null) {
 
 						definition.sortReferences();
 						definition.prepare(analyzer);
-						verifyVersion(analyzer, definition);
+						verifyVersion(analyzer, definition, packageVersion);
 
 						String name = "OSGI-INF/"
 								+ analyzer.validResourcePath(definition.name, "Invalid component name") + ".xml";
 						names.add(name);
 						analyzer.getJar().putResource(name, new TagResource(definition.getTag()));
+
+						if (definition.service != null) {
+							String[] objectClass = new String[definition.service.length];
+
+							for (int i = 0; i < definition.service.length; i++) {
+								Descriptors.TypeRef tr = definition.service[i];
+								objectClass[i] = tr.getFQN();
+							}
+							Arrays.sort(objectClass);
+							addProvidesHeader(objectClass, provides);
+						}
+						for (ReferenceDef ref : definition.references.values()) {
+							String objectClass = ref.service;
+							addRequireHeader(objectClass, requires);
+						}
+						maxVersion = definition.max(maxVersion, definition.version);
 					}
 				}
 			}
 		}
+		if (flags.contains(Flag.extender)
+				|| (packageVersion != null && packageVersion.compareTo(AnnotationReader.V1_3) >= 0)
+				|| maxVersion.compareTo(AnnotationReader.V1_3) >= 0) {
+			addExtenRequireHeader(requires);
+		}
 		sc = Processor.append(names.toArray(new String[names.size()]));
 		analyzer.setProperty(Constants.SERVICE_COMPONENT, sc);
+		updateHeader(analyzer, Constants.REQUIRE_CAPABILITY, requires);
+		updateHeader(analyzer, Constants.PROVIDE_CAPABILITY, provides);
 		return false;
 	}
 
 	/**
 	 * Verify that the component definition has a version that is <= than the
 	 * version of the component package that we build against.
+	 * 
+	 * @param version
+	 *            package version of org.osgi.service.component exported in
+	 *            build environment.
 	 */
-	private void verifyVersion(Analyzer analyzer, ComponentDef definition) throws Exception {
-		PackageRef component = analyzer.getPackageRef("org.osgi.service.component");
+	private void verifyVersion(Analyzer analyzer, ComponentDef definition, Version v) throws Exception {
+		if (v != null) {
+			if (definition.version.compareTo(v) > 0) {
+				SetLocation error = analyzer
+						.error("Generating XML for %s in type %s that uses a namespace version %s while you are building against %s",
```

Comment:
Why is this an error? This does not seem to be an error to me. Just because an older version of component is on the classpath should not fail my build. I am not even sure this is a warning. I might have one component which uses the older component package and another which is POJO but needs newer DS features.